### PR TITLE
ARROW-16794: [CI][C++][MinGW] Make CI jobs more stable

### PR DESCRIPTION
**Pull request by kou**
This pull request includes the followings:

* Build error fixes on debug build
  * Debug build was tried but it's slower than release build
  * So we don't change build type in this pull request
* Increase timeout to make MinGW related CI jobs buildable without cache
  * 60 minutes are short for MinGW related CI jobs without cache

### CODE REVIEW

**Comment by kou** [comment]
Debug build isn't fast...

**Comment by pitrou** [comment]
Perhaps disable CMAKE_UNITY_BUILD? (perhaps the workers lack RAM)

**Comment by kou** [comment]
`CMAKE_UNITY_BUILD` was unrelated... (CI jobs still timed out.)

It seems that we need to accept 60+ min build time without cache. We can build them in 15-25 min with cache. Can we increase `timeout-minutes` in `.github/workflows/cpp.yml` to `90` from `60`?

File: .github/workflows/cpp.yml [hunk 1]
```diff
@@ -260,7 +260,7 @@ jobs:
     name: AMD64 Windows MinGW ${{ matrix.mingw-n-bits }} C++
     runs-on: windows-2019
     if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
-    timeout-minutes: 60
+    timeout-minutes: 90
```
Comment (assignUser) [inline .github/workflows/cpp.yml:263]: Seeing that it also times out with 90: https://github.com/apache/arrow/runs/6912383497?check_suite_focus=true
Should we bump it up some more? A long build is annoying but a canceled build is just wasted time...

Comment (pitrou) [reply .github/workflows/cpp.yml:263]: If it times out with 90 then something is going really wrong.

**Comment by kou** [comment]
It seems that the following configurations are the best for now:

* Increase timeout to 90 minutes from 60 minutes to avoid time-out without cache
* Use release build not debug build
  * Release build is faster than debug build
  
If there are no objections, I'll merge this in a few days.

**Comment by wjones127** [comment]
Thank you @kou!

**Pull request merged**